### PR TITLE
feat: initial cross-platform shebang support

### LINKED ISSUE #23: Ability to execute files with shebang?
For example, say you had:

```json
{
  "tasks": {
    "my_task": "./my_script.js"
  }
}
```

And the script had:

```js
#!/usr/bin/env -S deno run --unstable --allow-write --allow-read --allow-run
// ...etc...
```

Perhaps if the script has a shebang starting with `#!/usr/bin/env -S` then deno task should just figure this out. This apparently already works on Mac so making this work on all platforms would reduce some issues.

From @jespertheend in https://github.com/denoland/deno/discussions/14036

> I like this method because it keeps the required permissions close to the actual file, and running the script is extremely simple because all you have to do is run ./tools/format.js.
>
> But the downside is... it doesn't work very well on Windows. Maybe I'm missing a simple solution for this, but as far as I can tell /usr/bin/env is not really a thing on Windows, and shebang support isn't great either. So Windows users are stuck with having to run scripts using deno run while trying to figure out what the required permissions are (or running with -A).
This wasn't a huge problem for me though. I'm mostly developing on macOS, so I was using shebangs for now, though I did want to have better cross platform support eventually.

### PR DESCRIPTION
**Pull request by dsherret**
Closes #23

### CODE REVIEW

File: src/shell/mod.rs [hunk 1]
```diff
@@ -684,7 +689,7 @@ fn execute_command_args(
   }
 }
 
-async fn evaluate_args(
+pub(crate) async fn evaluate_args(
```
Comment (dsherret) [inline src/shell/mod.rs:692]: Slightly messy...

Comment (dsherret) [reply src/shell/mod.rs:692]: I'll land a follow-up PR after this one to clean this up.

File: src/shell/commands/executable.rs [hunk 2]
```diff
@@ -1,3 +1,8 @@
+use std::borrow::Cow;
```
Comment (bartlomieju) [inline src/shell/commands/executable.rs:1]: Nitpick: missing copyright

File: src/shell/commands/executable.rs [hunk 3]
```diff
@@ -98,13 +118,167 @@ impl ShellCommand for ExecutableCommand {
...
+      Ok(std::env::current_exe()?)
+    }) {
+      Ok(command_path) => command_path,
+      Err(err) => return Err(err.into()),
+    };
+
+  // only bother checking for a shebang when the path has a slash
+  // in it because for global commands someone on Windows likely
+  // won't have a script with a shebang in it on Windows
+  if command_name.name.chars().any(|c| matches!(c, '\\' | '/')) {
```
Comment (bartlomieju) [inline src/shell/commands/executable.rs:172]: Oh I didn't know once can use backslash in shebags

File: src/shell/commands/executable.rs [hunk 4]
```diff
@@ -98,13 +118,167 @@ impl ShellCommand for ExecutableCommand {
...
+    args: Cow::Borrowed(original_args),
+  });
+}
+
+async fn parse_shebang_args(
+  text: &str,
+  context: &ShellCommandContext,
+) -> Result<Vec<String>> {
+  fn err_unsupported() -> Result<Vec<String>> {
+    anyhow::bail!("unsupported shebang. Please report this as a bug.")
```
Comment (bartlomieju) [inline src/shell/commands/executable.rs:210]: Print the shebang and add a link to issue tracker?

**Review by bartlomieju — APPROVED** [review]
Can you please open a PR to https://github.com/denoland/deno-docs that explains this in a few sentences?

File: src/shell/commands/executable.rs (hunk 3 above)
Comment (dsherret) [reply src/shell/commands/executable.rs:172]: I don't think you can. I might as well remove that.

**Comment by dsherret** [comment]
Will do when merging to CLI.

**Pull request merged**